// JSON [de]serialization for protobuf + embedded HTTP server and client in C++.
// Copyright (C) 2014 http://protojson.com/
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, write to the Free Software Foundation, Inc.,
// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

package pjcore;

import "pjcore/json.proto";

enum HttpStatusCode {
  HTTP_STATUS_CODE_CONTINUE = 100;
  HTTP_STATUS_CODE_SWITCHING_PROTOCOLS = 101;
  HTTP_STATUS_CODE_OK = 200;
  HTTP_STATUS_CODE_CREATED = 201;
  HTTP_STATUS_CODE_ACCEPTED = 202;
  HTTP_STATUS_CODE_NON_AUTHORITATIVE_INFORMATION = 203;
  HTTP_STATUS_CODE_NO_CONTENT = 204;
  HTTP_STATUS_CODE_RESET_CONTENT = 205;
  HTTP_STATUS_CODE_PARTIAL_CONTENT = 206;
  HTTP_STATUS_CODE_MULTIPLE_CHOICES = 300;
  HTTP_STATUS_CODE_MOVED_PERMANENTLY = 301;
  HTTP_STATUS_CODE_FOUND = 302;
  HTTP_STATUS_CODE_SEE_OTHER = 303;
  HTTP_STATUS_CODE_NOT_MODIFIED = 304;
  HTTP_STATUS_CODE_USE_PROXY = 305;
  HTTP_STATUS_CODE__UNUSED_ = 306;
  HTTP_STATUS_CODE_TEMPORARY_REDIRECT = 307;
  HTTP_STATUS_CODE_BAD_REQUEST = 400;
  HTTP_STATUS_CODE_UNAUTHORIZED = 401;
  HTTP_STATUS_CODE_PAYMENT_REQUIRED = 402;
  HTTP_STATUS_CODE_FORBIDDEN = 403;
  HTTP_STATUS_CODE_NOT_FOUND = 404;
  HTTP_STATUS_CODE_METHOD_NOT_ALLOWED = 405;
  HTTP_STATUS_CODE_NOT_ACCEPTABLE = 406;
  HTTP_STATUS_CODE_PROXY_AUTHENTICATION_REQUIRED = 407;
  HTTP_STATUS_CODE_REQUEST_TIMEOUT = 408;
  HTTP_STATUS_CODE_CONFLICT = 409;
  HTTP_STATUS_CODE_GONE = 410;
  HTTP_STATUS_CODE_LENGTH_REQUIRED = 411;
  HTTP_STATUS_CODE_PRECONDITION_FAILED = 412;
  HTTP_STATUS_CODE_REQUEST_ENTITY_TOO_LARGE = 413;
  HTTP_STATUS_CODE_REQUEST_URI_TOO_LONG = 414;
  HTTP_STATUS_CODE_UNSUPPORTED_MEDIA_TYPE = 415;
  HTTP_STATUS_CODE_REQUESTED_RANGE_NOT_SATISFIABLE = 416;
  HTTP_STATUS_CODE_EXPECTATION_FAILED = 417;
  HTTP_STATUS_CODE_INTERNAL_SERVER_ERROR = 500;
  HTTP_STATUS_CODE_NOT_IMPLEMENTED = 501;
  HTTP_STATUS_CODE_BAD_GATEWAY = 502;
  HTTP_STATUS_CODE_SERVICE_UNAVAILABLE = 503;
  HTTP_STATUS_CODE_GATEWAY_TIMEOUT = 504;
  HTTP_STATUS_CODE_HTTP_VERSION_NOT_SUPPORTED = 505;
}

message TextLocation {
  optional uint64 offset = 1;
  optional uint64 line = 2 [default = 1];
  optional uint64 column = 3 [default = 1];
}

message Error {
  optional string description = 1;
  optional string source_file = 2;
  optional int32 source_line = 3;
  optional int32 system_errno = 4;
  optional string system_errno_description = 5;
  optional int32 uv_errno = 6;
  optional string uv_errno_description = 7;
  optional int32 http_errno = 8;
  optional string http_errno_description = 9;
  optional HttpStatusCode http_status_code = 10;
  optional string http_reason_phrase = 11;
  optional string json_path = 12;
  optional TextLocation text_location = 13;
  repeated JsonValue.Property object_properties = 14;
  optional Error cause = 15;
}
